plugins/arm_thumb/thumb.ml

```diff
@@ -0,0 +1,209 @@
+    | `tSTMIA_UPD, dest :: _dest :: _unknown :: _nil_reg :: src_list ->
+      Mem.store_multiple dest src_list |> move
+    | `tLDMIA, dest :: _unknown :: _nil_reg :: src_list (* same as stmia *)
+    | `tLDMIA_UPD, dest :: _unknown :: _nil_reg :: src_list ->
+      Mem.load_multiple dest src_list |> move
+    | `tPUSH, _unknown :: _nil_reg :: src_list ->
+      Mem.push_multiple src_list |> move
+    | `tPOP, _unknown :: _nil_reg :: src_list ->
+      let has_pc = List.exists src_list 
+          (fun s -> match s with
+             |`Reg `PC -> true
+             | _ -> false) in
+      let pop_eff = Mem.pop_multiple src_list in
+      if has_pc then
+        ctrl (jmp (var Env.pc)) pop_eff addr
+      else move pop_eff
+    | _ -> move pass
+
+  let lift_bits insn ops =
+    let open Bits in
+    match insn, ops with
+    | `tSXTB, [|dest; src; _unknown; _|] -> sxtb dest src
+    | `tSXTH, [|dest; src; _unknown; _|] -> sxth dest src
+    | `tUXTB, [|dest; src; _unknown; _|] -> uxtb dest src
+    | `tUXTH, [|dest; src; _unknown; _|] -> uxth dest src
+    | _ -> []
+
+  (* these are not entirely complete *)
+  let lift_branch insn ops addr =
+    let open Defs in
+    let open Branch in
+    let addr = Core.int Env.value addr in
+    match insn, ops with
+    | `tBcc, [|target; cond; _cpsr|] -> tbcc cond target addr
+    | `tB, [|target; _unknown; _|] -> tb target addr
+    | `tBL, [|_unknown; _nil_reg; target; _cpsr|] -> tbl target
+    | `tBLXi, [|_unknown; _nil_reg; target; _cpsr|] -> tblxi target
+    | `tBLXr, [|_unknown; _nil_reg; target|] -> tblxr target
+    | `tBX, [|target; _unknown; _|] -> tbx target
+    | _ -> (skip, pass)
+
+  let lift_with (addr : Bitvec.t) (insn : Thumb_defs.insn)
+      (ops : Thumb_defs.op array) =
+    match insn with
+    | #move_insn -> lift_move insn ops |> DSL.expand |> move
+    | #mem_insn -> lift_mem insn ops addr
+    | #bits_insn -> lift_bits insn ops |> DSL.expand |> move
+    | #branch_insn -> 
+      let ctrl_eff, data_eff = lift_branch insn ops addr in
+      ctrl ctrl_eff data_eff addr (* var Env.pc *)
+
+end
+
+open Bap.Std
+
+(* this is a temporary fix since bap-mc disassembler couldn't recognize CMN *)
+let fix_cmnz insn mem = match insn with
+  | None -> if Memory.length mem = 2 then
+      let insn_word = Or_error.(ok (Size.of_int 16 >>= fun hw ->
+                                    Memory.get ~scale:hw mem >>= Word.extract ~hi:15 ~lo:6)) in
+      Option.(insn_word >>= fun insn -> 
+              if Word.equal (Word.of_int 10 0x10b) insn 
+              then Some `tCMNz 
+              else None)
+    else None
+  | Some insn -> Some insn
+
+let run_lifter _label addr insn mem 
+    (lifter : Bitvec.t -> Defs.insn -> Defs.op array -> unit Theory.eff) =
+  match fix_cmnz (Insns.of_basic insn) mem with
+  | None -> raise (Defs.Lift_Error "unknown instruction")
```

Comment:
Please, no exceptions, a lifter shall never fail - any input sequence of bytes is valid input. 
```suggestion
  | None -> KB.return Insn.empty
```

Reply:
this still needs to be addressed

Reply:
We do have a `try_with` at the place which run_lifter is called, I think this should be enough to address the underlying problem
```ocaml
    (match Or_error.try_with (
         fun () -> run_lifter label addr insn mem Lifter.lift_with
       ) with
```